### PR TITLE
Aiohttp server instrumentation

### PR DESCRIPTION
**Pull request by dmanchon**
# Description

Instrument aiohttp server. Right now there is only instrumentation for the client.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+async def middleware(request, handler):
+    """Middleware for aiohttp implementing tracing logic"""
+    if (
+        context.get_value("suppress_instrumentation")
+        or context.get_value(_SUPPRESS_HTTP_INSTRUMENTATION_KEY)
+        or _excluded_urls.url_disabled(request.url)
+    ):
+        return await handler(request)
+
+    token = context.attach(extract(request, getter=getter))
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py:145]: `context.attach` here is incorrect. The request being handled is from an external service, and the context contained within that request is a representation of the state of the external service. It doesn't make sense to set the external service's context as the current context for the service running this middleware.

The context should still be extracted from the request and supplied to `tracer.start_as_current_span`, but the `attach` and `detach` can be removed.

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    span_name, additional_attributes = get_default_span_details(request)
+
+    with tracer.start_as_current_span(
+        span_name,
+        kind=trace.SpanKind.SERVER,
+    ) as span:
+        if span.is_recording():
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py:152]: I think this `if` statement should be removed. In the case that this is a non-recording span, `set_attribute` is a no-op. Additionally, this middleware should _always_ call the request handler (line 158).

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            attributes = collect_request_attributes(request)
+            attributes.update(additional_attributes)
+            for key, value in attributes.items():
+                span.set_attribute(key, value)
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py:156]: Replace with `span.setattributes(attributes)`.

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            try:
+                resp = await handler(request)
+                set_status_code(span, resp.status)
+            finally:
+                context.detach(token)
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py:161]:
```suggestion
            resp = await handler(request)
            set_status_code(span, resp.status)
```

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+getter = AiohttpGetter()
+
+
+@web.middleware
+async def middleware(request, handler):
+    """Middleware for aiohttp implementing tracing logic"""
+    if (
+        context.get_value("suppress_instrumentation")
+        or context.get_value(_SUPPRESS_HTTP_INSTRUMENTATION_KEY)
+        or _excluded_urls.url_disabled(request.url)
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py:141]: `request.url` is not a `str`. This should be `request.url.path`.

**Comment by bmbouter** [comment]
@dmanchon I'm interested in testing and maybe helping to finish this PR? How can I test this PR after it's checked out? I have an aiohttp based app to test with. Also what do you know that needs finishing? The reviewer comments look all pretty straightforward, so after that what else is left?

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/setup.cfg [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+[options]
+python_requires = >=3.6
+package_dir=
+    =src
+packages=find_namespace:
+install_requires =
+    opentelemetry-api ~= 1.3
+    opentelemetry-semantic-conventions == 0.28b1
+    opentelemetry-instrumentation == 0.28b1
+    opentelemetry-util-http == 0.28b1
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/setup.cfg:46]: These requirements should be updated to the latest version

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/setup.cfg [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+[options.packages.find]
+where = src
+
+[options.extras_require]
+test =
+    opentelemetry-test-utils == 0.28b1
```
Comment (kennytrytek-wf) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/setup.cfg:53]: Update to latest version

**Comment by dralley** [comment]
> I remember one of my concerns was that aiohttp is both server and client, given that client already exists i wonder if this should live in the same module as the client or as an independent module, i guess if we want autoinstrumentation based on the installed packages it will end up installing both server and client modules.

Both server and client in one package is probably the best option by virtue of being simple + not real upsides of separating it.  It's ultimately not that much extra code.

**Comment by decko** [comment]
@dmanchon correct me if I'm wrong, but I didn't see any code for metrics. Is that right? Would be interesting to have it in this PR?
Also, would be possible for contributing to this PR specifically? Or we need to open a PR to your fork for then to me merged into this one?

**Comment by dralley** [comment]
@decko I would just open a new PR from your own branch which has these commits

**Comment by decko** [comment]
> @decko I would just open a new PR from your own branch which has these commits

What if we open it if we don't get a reply in two days?
In any case, I'm continuing the work from where @dmanchon left.

File: instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+    span_name, additional_attributes = get_default_span_details(request)
+
+    with tracer.start_as_current_span(
+        span_name,
+        kind=trace.SpanKind.SERVER,
+    ) as span:
+        attributes = collect_request_attributes(request)
+        attributes.update(additional_attributes)
+        span.setattributes(attributes)
```
Comment (aleuk0) [inline instrumentation/opentelemetry-instrumentation-aiohttp-server/src/opentelemetry/instrumentation/aiohttp_server/__init__.py:153]: little misspelling 

```suggestion
        span.set_attributes(attributes)
```

**Comment by lzchen** [comment]
Duplicate of https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1800

**Pull request closed without merging**